noxfile.py

```diff
@@ -52,9 +54,43 @@ def benchmark_model(session: nox.Session, subpackage: ModelSubpackagePath) -> No
     """Run pytest benchmarks for selected icon4py model subpackages."""
     _install_session_venv(session, extras=["dace", "io", "testing"], groups=["test"])
 
+    results_json_path = os.path.abspath(f"results_{subpackage.replace('/', '_')}.json")
     with session.chdir(f"model/{subpackage}"):
-        session.run(*"pytest -sv --benchmark-only".split(), *session.posargs)
-
+        session.run(*f"pytest \
+                       -v \
+                       --benchmark-only \
+                       --benchmark-warmup=on \
+                       --benchmark-warmup-iterations=30 \
+                       --benchmark-json={results_json_path}".split(),
+                       *session.posargs)
+
+@nox.session(python=["3.10", "3.11"], requires=["benchmark_model-{python}" + f"({subpackage.id})" for subpackage in MODEL_SUBPACKAGE_PATHS])
+def bencher_baseline(session: nox.Session) -> None:
+    """Run pytest benchmarks and upload them using Bencher (https://bencher.dev/) (cloud or self-hosted)."""
+    bencher_json_file_name = f"merged_benchmark_results_{session.python}.json"
+    merged_results = {"benchmarks": []}
+    files = glob.glob("results_*.json")
+    for file in files:
+        with open(file, "r") as f:
+            data = json.load(f)
+            merged_results["benchmarks"].extend(data["benchmarks"])
+    with open(bencher_json_file_name, "w") as f:
+        json.dump(merged_results, f, indent=4)
+
+    session.run(*f"bencher run \
+                   --project {os.environ['BENCHER_PROJECT']} \
+                   --token {os.environ['BENCHER_API_TOKEN']} \
+                   --branch main \
```

Comment:
Shouldn't the branch value be read directly from git? If we want to make sure that we only submit results to bencher from the `main` branch, maybe we should check at the beginning of the session, before running the tests, wether the currently checked out branch is `main`.

Reply:
The problem is that in the CI the folder does not contain the .git folder. So we cannot get any info for the repo, except for accessing the CI env vars. In any case, this pipeline is supposed to be executed automatically only on the main branch. I would complicate it more for now.